dali/pipeline/operators/detection/box_encoder.cc

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+
+#include "dali/pipeline/operators/detection/box_encoder.h"
+
+namespace dali {
+vector<float> BoxEncoder<CPUBackend>::CalculateIous(vector<BoundingBox> boxes) {
+  vector<float> ious(boxes.size() * NumAnchors());
+
+  for (auto bbox_idx = 0u; bbox_idx < boxes.size(); ++bbox_idx) {
+    int best_idx = -1;
+    float best_iou = -1.;
+
+    for (int anchor_idx = 0; anchor_idx < NumAnchors(); ++anchor_idx) {
+      ious[bbox_idx * anchors_.size() + anchor_idx] =
+          boxes[bbox_idx].IntersectionOverUnion(anchors_[anchor_idx]);
+
+      if (ious[bbox_idx * anchors_.size() + anchor_idx] >= best_iou) {
+        best_iou = ious[bbox_idx * NumAnchors() + anchor_idx];
+        best_idx = anchor_idx;
+      }
+    }
+
+    // For best default box matched with current object let iou = 2, to make sure there is a match,
+    // as this object will be the best (highest IOU), for this default box
+    ious[bbox_idx * NumAnchors() + best_idx] = 2.;
+  }
+
+  return ious;
+}
+
+void BoxEncoder<CPUBackend>::MatchBoxesWithAnchors(
```

Comment:
This function is doing several things, would be good to break this down a little bit

Reply:
Fixed by extracting parts of this method as separate methods.

Reply:
Could we make thsi function return the matched boxes and anchors, and perform the write on a separate function? Currently is a bit confusing that the write operation is buried at the end of a call that is supposed to match data